### PR TITLE
Add coordinator logic

### PR DESCRIPTION
**Pull request by r-marques**
This is still a work in progress but the main goal is to have the coordinator logic independent of gRPC. This makes it easier to unit tests the coordinator logic and makes it easier to use without gRPC.

The coordinator logic is basically just one method that implements a state machine that reacts to gRPC messages passed to it. This messages can be used both over gRPC or locally without networking.
- Added [intersphinx](http://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) to link to external documentation

### CODE REVIEW

File: xain/grpc/coordinator.py [hunk 1]
```diff
@@ -113,102 +120,144 @@ def monitor_heartbeats(participants, terminate_event):
...
+                    f" # participants: {self.participants.len()}"
+                )
+
+            # Change the state to ROUND if we are in STANDBY and already have enough participants
+            if (
+                self.state == STANDBY
+                and self.participants.len() == self.required_participants
+            ):
+                # TODO: We may need to make this update thread safe
+                self.state = ROUND
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]: maybe change state in the `if` block above (where we should be guaranteed to be in STANDBY)? If we already have `required_participants` we shouldn't need to check this.

File: xain/grpc/coordinator.py [hunk 2]
```diff
@@ -113,102 +120,144 @@ def monitor_heartbeats(participants, terminate_event):
...
+        # local model
+        self.theta_updates = []
+        self.histories = []
+        self.metricss = []
+
+        # state variables
+        self.state = STANDBY
+        self.round = 0
+
+    def on_message(self, message, peer_id):
```
[the remark this answers is not in the document]
Comment (r-marques) [reply xain/grpc/coordinator.py]: No particular reason. Just made it easier to reason about the state machine. I will probably split into multiple functions.

File: xain/grpc/test_coordinator_logic.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    coordinator.on_message(coordinator_pb2.EndTrainingRequest(), "peer1")
+
+    assert len(coordinator.theta_updates) == 1
+    assert len(coordinator.histories) == 1
+    assert len(coordinator.metricss) == 1
+
+
+def test_end_training_round_update():
+    # Test that the round number is updated once all participants sent their updates
+    coordinator = Coordinator(required_participants=1)
```
Comment (finiteprods) [inline xain/grpc/test_coordinator_logic.py]: Perhaps you can also test the more interesting case with 2 peers

File: xain/grpc/test_coordinator_logic.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    coordinator.on_message(coordinator_pb2.EndTrainingRequest(), "peer1")
+
+    # After one participant sends its updates we should have one update in the coordinator
+    assert len(coordinator.theta_updates) == 1
+    assert len(coordinator.histories) == 1
+    assert len(coordinator.metricss) == 1
+
+    coordinator.on_message(coordinator_pb2.EndTrainingRequest(), "peer2")
+
+    # once the second participant delivers its updates we the round ends and the local models
```
Comment (finiteprods) [inline xain/grpc/test_coordinator_logic.py]:
```suggestion
    # once the second participant delivers its updates the round ends and the local models
```

File: xain/grpc/coordinator.py [hunk 5]
```diff
@@ -20,55 +27,58 @@ class ParticipantContext:
...
     In the future we may store more information like in what state a participant is in e.g.
     IDLE, RUNNING, ...
+
+    Args:
+        participant_id (:obj:`str`): The id of the participant. Typically a
+            host:port or public key when using SSL.
     """
 
-    def __init__(self, participant_id):
+    def __init__(self, participant_id: str) -> None:
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]: Very minor comment, generally in this file: I think we don't need to mention the type again in the docstring if we have the `mypy` annotation (benefits in being in one place, etc.)

File: xain/grpc/coordinator.py [hunk 6]
```diff
@@ -103,112 +113,336 @@ def update_expires(self, participant_id):
...
+    The states of the Coordinator are:
+        STANDBY: The coordinator is in standby mode, typically when waiting for
+        participants to connect. In this mode the only messages that the
+        coordinator can receive are :class:`~.coordinator_pb2.RendezvousRequest`
+        and :class:`~.coordinator_pb2.HeartbeatRequest`.
+
+        ROUND: A round is currently in progress. During a round the only
+        messages the coordinator can receive are
+        :class:`~.coordinator_pb2.StartTrainingRequest` and
+        :class:`~.coordinator_pb2.EndTrainingRequest`.
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]: Is this right? Aren't heartbeat requests still serviced? And rendezvous' can also be told to try `LATER`?

File: xain/grpc/coordinator.py [hunk 7]
```diff
@@ -103,112 +113,336 @@ def update_expires(self, participant_id):
...
+        num_rounds (:obj:`int`, optional): The number of rounds of the training
+            session. Defaults to 10.
+        required_participants(:obj:`int`, optional): The minimum number of
+            participants required to perform a round. Defaults to 10.
+        aggregator: (:class:`~.Aggregator`, optional): The type of aggregation
+            to perform at the end of each round. Defaults to
+            :class:`~.FederatedAveragingAgg`.
+        theta (:obj:`list` of :class:`~numpy.ndarray`, optional): The weights of
+            the global model. Defaults to [].
+        epochs (:obj:`int`, optional): TODO.  Defaults to 0.
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]: I can suggest a description (after I double-check it myself) if you like

File: xain/grpc/coordinator.py [hunk 8]
```diff
@@ -103,112 +113,336 @@ def update_expires(self, participant_id):
...
+        epochs_base (:obj:`int`, optional): TODO. Defautls to 0.
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]:
```suggestion
        epochs_base (:obj:`int`, optional): TODO. Defaults to 0.
```

File: xain/grpc/coordinator.py [hunk 9]
```diff
@@ -103,112 +113,336 @@ def update_expires(self, participant_id):
...
+        self.coordinator = coordinator
+
+    def Rendezvous(
+        self, request: coordinator_pb2.RendezvousRequest, context: grpc.ServicerContext
+    ) -> coordinator_pb2.RendezvousReply:
+        """The Rendezvous gRPC method.
+
+        A participant contacts the coordinator and the coordinator adds the participant to
+        its list of participants. If the coordinator already has all the participants it
+        needs it tells the participant to try again later.
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]:
```suggestion
        tells the participant to try again later.
```

File: xain/grpc/coordinator.py [hunk 10]
```diff
@@ -103,112 +113,336 @@ def update_expires(self, participant_id):
...
+        Args:
+            request (:class:`~.coordinator_pb2.EndTrainingRequest`): The
+                participant's request. The request contains the updated weights as
+                a result of the training as well as any metrics helpful for the
+                :class:`~.Coordinator`.
+            context (:class:`~grpc.ServicerContext`): The context associated with the gRPC request.
+
+        Returns:
+            :class:`~.coordinator_pb2.EndTrainingReply`: The reply to the
+            participant's request. The reply is just and acknowledgment that
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]:
```suggestion
            participant's request. The reply is just an acknowledgment that
```

File: xain/grpc/coordinator.py [hunk 11]
```diff
@@ -103,112 +113,336 @@ def update_expires(self, participant_id):
...
     Note:
-        This is meant to be run inside a thread and expects a `threading.Event` to know when
-        it should terminate.
+        This is meant to be run inside a thread and expects an
+        :class:`~threading.Event`, to know when it should terminate.
 
     Args:
-        participants (Participants): The participants to monitor.
+        coordinator (:class:`~.Coordinator`): The coordinator to monitor for heartbeats.
+        terminate_event (:class:`~threading.Event`): A threading even to signal
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]:
```suggestion
        terminate_event (:class:`~threading.Event`): A threading event to signal
```

File: xain/grpc/coordinator.py [hunk 12]
```diff
@@ -103,112 +113,336 @@ def update_expires(self, participant_id):
...
-        self.theta_updates.append(theta_update)
-        self.histories.append({k: list(hv.values) for k, hv in his.items()})
-        self.metricss.append((cid, list(vbc)))
-        # reply
-        return coordinator_pb2.EndTrainingReply()
-
-
-def serve():
-    participants = Participants()
+    This methods just creates the :class:`~.Coordinator`, setups all threading
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]:
```suggestion
    This methods just creates the :class:`~.Coordinator`, sets up all threading
```

File: xain/grpc/test_coordinator_logic.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    result = coordinator.on_message(coordinator_pb2.StartTrainingRequest(), "peer1")
+    received_theta = [proto_to_ndarray(nda) for nda in result.theta]
+
+    np.testing.assert_equal(test_theta, received_theta)
+
+
+def test_end_training():
+    coordinator = Coordinator(required_participants=2)
+    coordinator.on_message(coordinator_pb2.RendezvousRequest(), "peer1")
+    coordinator.on_message(coordinator_pb2.RendezvousRequest(), "peer2")
```
[the remark this answers is not in the document]
Comment (r-marques) [reply xain/grpc/test_coordinator_logic.py]: In order to check the status of the updates I need to do it during a round because at the end of the round the local state of the participants is reset. If I had only 1 participant the round would finish after the first EndTrainingRequest and I wouldn't be able to do the asserts.

File: xain/grpc/coordinator.py (hunk 5 above)
Comment (r-marques) [reply xain/grpc/coordinator.py]: I don't know exactly how sphinx deals with that. Specially if using intersphinx to link to external documentation

File: xain/grpc/coordinator.py (hunk 6 above)
Comment (r-marques) [reply xain/grpc/coordinator.py]: That is true but I just wanted to make the documentation as simple as possible.

I left a note https://github.com/xainag/xain/pull/128/files#diff-6fad7f1e8d3e49838ad5b60a5cd57f52R143

File: xain/grpc/coordinator.py (hunk 7 above)
Comment (r-marques) [reply xain/grpc/coordinator.py]: sure :+1:

File: xain/grpc/coordinator.py (hunk 5 above)
Comment (finiteprods) [reply xain/grpc/coordinator.py]: It would be good if sphinx were smart enough with `mypy` annotations but I don't know. Anyway, this is a separate discussion and certainly no blocker here.

File: xain/grpc/coordinator.py (hunk 6 above)
Comment (finiteprods) [reply xain/grpc/coordinator.py]: I would suggest rewording "...the **only** messages..." to "...the **important** messages..." for clarity.

File: xain/grpc/coordinator.py (hunk 7 above)
Comment (finiteprods) [reply xain/grpc/coordinator.py]:
```suggestion
        epochs (:obj:`int`, optional): The number of rounds of training, internal to the Participant.  Defaults to 0.
```
(wrap the long line over where appropriate)

File: xain/grpc/coordinator.py (hunk 8 above)
Comment (finiteprods) [reply xain/grpc/coordinator.py]:
```suggestion
        epochs_base (:obj:`int`, optional): Global number of epochs since last round. Defaults to 0.
```

File: xain/grpc/test_coordinator_logic.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (finiteprods) [reply xain/grpc/test_coordinator_logic.py]: Yes that's clear now, but not obvious from reading the test. A comment, perhaps?

File: xain/grpc/coordinator.py [hunk 14]
```diff
@@ -155,7 +155,7 @@ class Coordinator:
         theta (:obj:`list` of :class:`~numpy.ndarray`, optional): The weights of
             the global model. Defaults to [].
         epochs (:obj:`int`, optional): TODO.  Defaults to 0.
-        epochs_base (:obj:`int`, optional): TODO. Defautls to 0.
+        epochs_base (:obj:`int`, optional): TODO. Defaults to 0.
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]: Description: 

> Global number of epochs since last round

File: xain/grpc/coordinator.py [hunk 15]
```diff
@@ -155,7 +155,7 @@ class Coordinator:
         theta (:obj:`list` of :class:`~numpy.ndarray`, optional): The weights of
             the global model. Defaults to [].
         epochs (:obj:`int`, optional): TODO.  Defaults to 0.
```
Comment (finiteprods) [inline xain/grpc/coordinator.py]: Description:

> Number of training iterations local to Participant

File: xain/grpc/test_coordinator_logic.py [hunk 16]
```diff
@@ -75,14 +78,16 @@ def test_end_training():
...
+    # Test that the round number is updated once all participants send their updates
+    coordinator = Coordinator(required_participants=2)
     coordinator.on_message(coordinator_pb2.RendezvousRequest(), "peer1")
+    coordinator.on_message(coordinator_pb2.RendezvousRequest(), "peer2")
 
     # check that we are currently in round 1
     assert coordinator.round == 1
 
     coordinator.on_message(coordinator_pb2.EndTrainingRequest(), "peer1")
+    coordinator.on_message(coordinator_pb2.EndTrainingRequest(), "peer2")
```
Comment (finiteprods) [inline xain/grpc/test_coordinator_logic.py]: Between here add an assert that we're still in round 1

Comment (r-marques) [reply xain/grpc/test_coordinator_logic.py]: Yes. The round is updated only after we get updates == required_participantes

File: xain/grpc/test_coordinator_logic.py [hunk 17]
```diff
@@ -66,6 +66,9 @@ def test_start_training():
 
 
 def test_end_training():
+    # we need to participants so that we can check the status of the local update mid round
```
Comment (finiteprods) [inline xain/grpc/test_coordinator_logic.py]: s/to/two

File: xain/grpc/test_coordinator_logic.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+def test_end_training_round_update():
+    # Test that the round number is updated once all participants send their updates
+    coordinator = Coordinator(required_participants=2)
+    coordinator.on_message(coordinator_pb2.RendezvousRequest(), "peer1")
+    coordinator.on_message(coordinator_pb2.RendezvousRequest(), "peer2")
+
+    # check that we are currently in round 1
+    assert coordinator.round == 1
+
+    coordinator.on_message(coordinator_pb2.EndTrainingRequest(), "peer1")
```
Comment (finiteprods) [inline xain/grpc/test_coordinator_logic.py]:
```suggestion
    coordinator.on_message(coordinator_pb2.EndTrainingRequest(), "peer1")
    # check we are still in round 1
    assert coordinator.round == 1
```

File: xain/grpc/test_coordinator_logic.py (hunk 16 above)
Comment (finiteprods) [reply xain/grpc/test_coordinator_logic.py]: so assert it? See my suggestion.

File: xain/grpc/test_coordinator_logic.py (hunk 17 above)
Comment (r-marques) [reply xain/grpc/test_coordinator_logic.py]: sorry I missed this suggestion the first time.
